test/parallel_api/algorithm/alg.nonmodifying/all_of.pass.cpp

```diff
@@ -111,7 +111,7 @@ main()
     test<std::int32_t>(8 * sizeof(std::int32_t));
     test<std::uint16_t>(8 * sizeof(std::uint16_t));
     test<float64_t>(53);
-    test<bool>(1);
+    test<unsigned char>(1);
```

Comment:
Actually, we are reducing the test coverage of oneDPL, especially very interesting case - std::vector<bool>. In the standard libraries there is a specialization of `std::vector` for `bool` type. Furthermore, it does not necessarily store its elements as a contiguous array. But, `std::vector<bool>::iterator` -is a random access iterator. And I don't see (so while) any reason to unsupport it. I think we should find as error in the oneDPL implementation and fix it.

Reply:
http://icl-jenkins2.sc.intel.com:8080/blue/organizations/jenkins/OneDPL_CI%2FRHEL_Check/detail/RHEL_Check/2861/pipeline :

[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/20220209/build/linux_prod/compiler/linux/bin-llvm/../include/sycl/CL/sycl/buffer.hpp:256:64: error: cannot take the address of an rvalue of type 'std::_Bit_const_iterator::const_reference' (aka 'bool')
[2022-02-16T06:29:51.010Z]     impl->constructorNotification(CodeLoc, (void *)impl.get(), &*first,
[2022-02-16T06:29:51.010Z]                                                                ^~~~~~~
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/include/oneapi/dpl/internal/../pstl/hetero/dpcpp/utils_ranges_sycl.h:328:48: note: in instantiation of function template specialization 'sycl::buffer<bool, 1, sycl::detail::aligned_allocator<char>, void>::buffer<std::_Bit_const_iterator, 1, void, void>' requested here
[2022-02-16T06:29:51.010Z]             _copy_direct_tag{}, [&]() { return sycl::buffer<val_t<_Iter>, 1>(__first, __last); },
[2022-02-16T06:29:51.010Z]                                                ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/include/oneapi/dpl/internal/../pstl/hetero/dpcpp/utils_ranges_sycl.h:536:20: note: in instantiation of function template specialization 'oneapi::dpl::__ranges::__get_sycl_range<sycl::access::mode::read, std::_Bit_const_iterator>::copy_direct<std::_Bit_const_iterator, std::integral_constant<bool, true>>' requested here
[2022-02-16T06:29:51.010Z]         auto buf = copy_direct(__first, __last, copy_direct_tag());
[2022-02-16T06:29:51.010Z]                    ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:575:18: note: in instantiation of function template specialization 'oneapi::dpl::__ranges::__get_sycl_range<sycl::access::mode::read, std::_Bit_const_iterator>::operator()<std::_Bit_const_iterator>' requested here
[2022-02-16T06:29:51.010Z]     auto __buf = __keep(__first, __last);
[2022-02-16T06:29:51.010Z]                  ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/include/oneapi/dpl/pstl/glue_algorithm_impl.h:46:37: note: in instantiation of function template specialization 'oneapi::dpl::__internal::__pattern_any_of<oneapi::dpl::execution::device_policy<oneapi::dpl::__par_backend_hetero::__unique_kernel_name<test_any_of<bool>, 1>> &, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>>' requested here
[2022-02-16T06:29:51.010Z]     return oneapi::dpl::__internal::__pattern_any_of(
[2022-02-16T06:29:51.010Z]                                     ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/parallel_api/algorithm/alg.nonmodifying/any_of.pass.cpp:42:31: note: in instantiation of function template specialization 'oneapi::dpl::any_of<oneapi::dpl::execution::device_policy<oneapi::dpl::__par_backend_hetero::__unique_kernel_name<test_any_of<bool>, 1>> &, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>>' requested here
[2022-02-16T06:29:51.010Z]         auto actualr = ::std::any_of(exec, begin, end, pred);
[2022-02-16T06:29:51.010Z]                               ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/support/iterator_utils.h:254:13: note: (skipping 1 context in backtrace; use -ftemplate-backtrace-limit=0 to see all)
[2022-02-16T06:29:51.010Z]             op(::std::forward<Rest>(rest)...);
[2022-02-16T06:29:51.010Z]             ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/support/iterator_utils.h:380:9: note: in instantiation of function template specialization 'TestUtils::invoke_if_<std::integral_constant<bool, false>, std::integral_constant<bool, true>>::operator()<test_any_of<bool>, oneapi::dpl::execution::device_policy<oneapi::dpl::__par_backend_hetero::__unique_kernel_name<test_any_of<bool>, 1>> &, std::_Bit_const_iterator, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>, bool>' requested here
[2022-02-16T06:29:51.010Z]         invoke_if<Iterator>()(::std::distance(inputBegin, inputEnd) <= sizeLimit, op, exec,
[2022-02-16T06:29:51.010Z]         ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/support/utils_sycl.h:157:9: note: in instantiation of function template specialization 'TestUtils::iterator_invoker<std::random_access_iterator_tag, std::integral_constant<bool, false>>::operator()<oneapi::dpl::execution::device_policy<oneapi::dpl::__par_backend_hetero::__unique_kernel_name<test_any_of<bool>, 1>>, test_any_of<bool>, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>, bool>' requested here
[2022-02-16T06:29:51.010Z]         iterator_invoker<::std::random_access_iterator_tag, /*IsReverse*/ ::std::false_type>()(
[2022-02-16T06:29:51.010Z]         ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/support/utils.h:715:9: note: in instantiation of function template specialization 'TestUtils::invoke_on_all_hetero_policies<1>::operator()<test_any_of<bool>, std::_Bit_const_iterator, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>, bool>' requested here
[2022-02-16T06:29:51.010Z]         invoke_on_all_hetero_policies<CallNumber>()(op, ::std::forward<T>(rest)...);
[2022-02-16T06:29:51.010Z]         ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/parallel_api/algorithm/alg.nonmodifying/any_of.pass.cpp:62:9: note: in instantiation of function template specialization 'TestUtils::invoke_on_all_policies<1>::operator()<test_any_of<bool>, std::_Bit_const_iterator, std::_Bit_const_iterator, TestUtils::is_equal_to<bool>, bool>' requested here
[2022-02-16T06:29:51.010Z]         invoke_on_all_policies<1>()(test_any_of<T>(), in.cbegin(), in.cend(), is_equal_to<T>(spike), false);
[2022-02-16T06:29:51.010Z]         ^
[2022-02-16T06:29:51.010Z] /export/users/sys_bbsycl/RHEL_Node_Inner/workspace/OneDPL_CI/RHEL_Check/src/test/parallel_api/algorithm/alg.nonmodifying/any_of.pass.cpp:100:5: note: in instantiation of function template specialization 'test<bool>' requested here
[2022-02-16T06:29:51.010Z]     test<bool>(1);
[2022-02-16T06:29:51.010Z]     ^
[2022-02-16T06:29:51.010Z] 1 error generated.

Reply:
From my point of view, code from SYCL library is not compiled. because in this code we trying to get address of iteradet value and this is unable to do for iterator of std::vector<bool>